broker.go

```diff
@@ -0,0 +1,706 @@
+		}
+	}
+}
+
+// leaderConnection returns connection to leader for given partition. If
+// connection does not exist, broker will try to connect first and add store
+// connection for any further use.
+// Failed connection retry is controlled by broker configuration.
+func (b *Broker) leaderConnection(topic string, partition int32) (conn *connection, err error) {
+	endpoint := fmt.Sprintf("%s:%d", topic, partition)
+
+	for retry := 0; retry < b.conf.LeaderRetryLimit; retry++ {
+		b.mu.Lock()
+
+		nodeID, ok := b.metadata.endpoints[endpoint]
+		if !ok {
+			if err := b.refreshMetadata(); err != nil {
+				b.conf.Log.Printf("cannot refresh metadata: %s", err)
+				b.mu.Unlock()
+				return nil, err
+			}
+			nodeID, ok = b.metadata.endpoints[endpoint]
+			if !ok {
+				b.mu.Unlock()
+				return nil, proto.ErrUnknownTopicOrPartition
+			}
+		}
+
+		conn, ok = b.conns[nodeID]
+		if !ok {
+			addr, ok := b.metadata.nodes[nodeID]
+			if !ok {
+				b.conf.Log.Printf("no information about node %d", nodeID)
+				b.mu.Unlock()
+				return nil, proto.ErrBrokerNotAvailable
+			}
+			conn, err = NewConnection(addr, b.conf.DialTimeout)
+			if err != nil {
+				b.conf.Log.Printf("cannot connect to node %s: %s", addr, err)
+				b.mu.Unlock()
+				time.Sleep(b.conf.LeaderRetryWait)
+				continue
+			}
+			b.conns[nodeID] = conn
+		}
+		b.mu.Unlock()
+		return conn, nil
+	}
+	return nil, err
+}
+
+// offset will return offset value for given partition. Use timems to specify
+// which offset value should be returned.
+func (b *Broker) offset(topic string, partition int32, timems int64) (offset int64, err error) {
+	conn, err := b.leaderConnection(topic, partition)
+	if err != nil {
+		return 0, err
+	}
+	resp, err := conn.Offset(&proto.OffsetReq{
+		ClientID:  b.conf.ClientID,
+		ReplicaID: -1, // any client
+		Topics: []proto.OffsetReqTopic{
+			proto.OffsetReqTopic{
+				Name: topic,
+				Partitions: []proto.OffsetReqPartition{
+					proto.OffsetReqPartition{
+						ID:         partition,
+						TimeMs:     timems,
+						MaxOffsets: 2,
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		return 0, err
+	}
+	found := false
+	for _, t := range resp.Topics {
+		if t.Name != topic {
+			b.conf.Log.Printf("unexpected topic information received: %s (expecting %s)", t.Name)
+			continue
+		}
+		for _, part := range t.Partitions {
+			if part.ID != partition {
+				b.conf.Log.Printf("unexpected partition information received: %s:%s (expecting %s)", t.Name, part.ID, partition)
+				continue
+			}
+			found = true
+			// happends when there are no messages
+			if len(part.Offsets) == 0 {
+				offset = 0
+			} else {
+				offset = part.Offsets[0]
+			}
+			err = part.Err
+		}
+	}
+	if !found {
+		return 0, errors.New("incomplete fetch response")
+	}
+	return offset, err
+}
+
+// OffsetEarliest returns offset of the oldest message available in given partition.
+func (b *Broker) OffsetEarliest(topic string, partition int32) (offset int64, err error) {
```

Comment:
You have done a decent job in abstracting away the crazy (or as the authors like to call "funky") api.